core/queueref_test.go

```diff
@@ -56,3 +61,130 @@ func TestQueueRef(t *testing.T) {
+	core.Put("key1", core.NewObj(val[0], -1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key1")
+	core.Put("key2", core.NewObj(val[1], -1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key2")
+	core.Put("key3", core.NewObj(val[2], -1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key3")
+	for i := 0; i < len(val); i++ {
+		qe, err := qr.Remove()
+		assert.Check(t, err == nil && qe.Obj.Value == val[i], fmt.Sprintf("removal from queue with multiple non expired keys failed, Expected : %d , Got : %d\n", val[i], qe.Obj.Value))
+	}
+}
+
+// Test for removing from queue with expired keys before non-expired
+func TestRemoveExpiredBeforeNonExpire(t *testing.T) {
+	qr := core.NewQueueRef()
+	val := [3]int{10, 20}
+	core.Put("key1", core.NewObj(val[0], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key1")
+	core.Put("key2", core.NewObj(val[1], -1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key2")
+	time.Sleep(2 * time.Millisecond)
+	qe, err := qr.Remove()
+	assert.Check(t, err == nil || qe.Obj.Value == val[1], fmt.Sprintf("test for removing expired key before non-expired failed , Expected : %d, Got : %d\n", val[1], qe.Obj.Value))
+}
+
+// Test for removing from queue with multiple expired keys before non-expired
+func TestRemoveMultipleExpiredBeforeNonExpire(t *testing.T) {
+	qr := core.NewQueueRef()
+	val := [3]int{10, 20, 30}
+	core.Put("key1", core.NewObj(val[0], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key1")
+	core.Put("key2", core.NewObj(val[1], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key2")
+	core.Put("key3", core.NewObj(val[2], -1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key3")
+	time.Sleep(2 * time.Millisecond)
+	qe, err := qr.Remove()
+	assert.Check(t, err == nil || qe.Obj.Value == val[2], fmt.Sprintf("test for removing mulitple expired key before non-expired failed , Expected : %d, Got %d\n", val[2], qe.Obj.Value))
+}
+
+// Test for removing from queue with all expired keys
+func TestRemoveAllExpired(t *testing.T) {
+	qr := core.NewQueueRef()
+	val := [3]int{10, 20, 30}
+	core.Put("key1", core.NewObj(val[0], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key1")
+	core.Put("key2", core.NewObj(val[1], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key2")
+	core.Put("key3", core.NewObj(val[2], 1, core.OBJ_TYPE_STRING, core.OBJ_ENCODING_INT))
+	qr.Insert("key3")
+	time.Sleep(2 * time.Millisecond)
+	// remove from empty queue
+	_, err := qr.Remove()
+	assert.Equal(t, err, core.ErrQueueEmpty, fmt.Sprintf("test for removing from empty queue failed Expected : %s, Got : %s\n", core.ErrQueueEmpty, err))
+}
+
+// Benchmark queueref by inserting expired, non-expired and expired keys in order and removing them
+func BenchmarkQueueRef(b *testing.B) {
+	benchmarkCases := []struct {
+		name         string
+		expiredFirst int
+		nonExpired   int
+		expiredLast  int
+	}{
+		{"Small", 10000, 20000, 30000},
```

Comment:
Trying to understand something. Should the `expiredLast` not be equal to `nonExpired`?
You are waiting for 2ms before you run the benchmark for qr.Remove(), wouldn't the `expiredFirst` keys have expired by then?

If my understanding is correct then it may be better to call the Remove() operation only `nonExpired` number of times, and finally verify that the queue is empty at the end of it (you can call b.StopTimer() before the assertion).

Almost maybe it would be better to rename the fields to expiredKeys, nonExpiredKeys.